table/tables/tables.go

```diff
@@ -1190,23 +1194,171 @@ func init() {
+			ok          bool
+		)
+		if seq.base == seq.end {
+			// There is no cache yet.
+			updateCache = true
+		} else {
+			// Seek the first valid value in cache.
+			offset = seq.getOffset()
+			if seq.meta.Increment > 0 {
+				nextVal, ok = autoid.SeekToFirstSequenceValue(seq.base, seq.meta.Increment, offset, seq.base, seq.end)
+			} else {
+				nextVal, ok = autoid.SeekToFirstSequenceValue(seq.base, seq.meta.Increment, offset, seq.end, seq.base)
+			}
+			if !ok {
+				updateCache = true
+			}
+		}
+		if !updateCache {
+			return nil
+		}
+		// Update batch alloc from kv storage.
+		sequenceAlloc, err1 := getSequenceAllocator(t.allocs)
+		if err1 != nil {
+			return err1
+		}
+		seq.base, seq.end, seq.round, err1 = sequenceAlloc.AllocSeqCache(t.tableID)
+		if err1 != nil {
+			return err1
+		}
+		// Seek the first valid value in new cache.
+		// Offset may have changed cause the round is updated.
+		offset = seq.getOffset()
+		if seq.meta.Increment > 0 {
+			nextVal, ok = autoid.SeekToFirstSequenceValue(seq.base, seq.meta.Increment, offset, seq.base, seq.end)
+		} else {
+			nextVal, ok = autoid.SeekToFirstSequenceValue(seq.base, seq.meta.Increment, offset, seq.end, seq.base)
+		}
+		if !ok {
+			return errors.New("can't find the first value in sequence cache")
+		}
+		return nil
+	}()
+	// Sequence alloc in kv store error.
+	if err != nil {
+		if err == autoid.ErrAutoincReadFailed {
+			return 0, table.ErrSequenceHasRunOut.GenWithStackByArgs(dbName, seqName)
+		}
+		return 0, err
+	}
+	seq.base = nextVal
+	return nextVal, nil
 }
 
 // SetSequenceVal implements util.SequenceTable SetSequenceVal interface.
+// The returned bool indicates the newVal is already under the base.
 func (t *TableCommon) SetSequenceVal(newVal int64) (int64, bool, error) {
-	if t.sequence < newVal {
-		t.sequence = newVal
+	seq := t.sequence
+	if seq == nil {
+		// TODO: refine the error.
+		return 0, false, errors.New("sequenceCommon is nil")
+	}
+	seq.mu.Lock()
+	defer seq.mu.Unlock()
+
+	if seq.meta.Increment > 0 {
+		if newVal <= t.sequence.base {
+			return 0, true, nil
+		}
+		if newVal <= t.sequence.end {
+			t.sequence.base = newVal
+			return newVal, false, nil
+		}
+	} else {
+		if newVal >= t.sequence.base {
+			return 0, true, nil
+		}
+		if newVal >= t.sequence.end {
+			t.sequence.base = newVal
+			return newVal, false, nil
+		}
+	}
+
+	// Invalid the current cache.
+	t.sequence.base = t.sequence.end
+
+	// Rebase from kv storage.
+	sequenceAlloc, err := getSequenceAllocator(t.allocs)
+	if err != nil {
+		return 0, false, err
+	}
+	err = sequenceAlloc.Rebase(t.tableID, newVal, false)
+	if err != nil {
+		return 0, false, err
 	}
-	// TODO: implement it with sequence rebase logic.
 	return newVal, false, nil
 }
 
+func (s *sequenceCommon) getOffset() int64 {
+	offset := s.meta.Start
+	if s.meta.Cycle && s.round > 0 {
+		if s.meta.Increment > 0 {
+			offset = s.meta.MinValue
+		} else {
+			offset = s.meta.MaxValue
+		}
+	}
+	return offset
+}
```

Comment:
Please use `mu.RLock()` to protect `s.round`.

Reply:
getOffset is only used in `GetSequenceNextVal` & `SetSequenceVal`, which are under the `mu.Lock()`.